core/query/sql.ml

```diff
@@ -137,138 +152,162 @@ let order_by_clause n =
+        let tables = List.map (fun (t, x) -> Format.asprintf "%a as %s" pp_quote t (string_of_table_var x)) tables in
+        pr_select ppf fields tables condition os
+    | Delete { del_table; del_where } ->
+        pr_delete ppf del_table del_where
+    | Update { upd_table; upd_fields; upd_where } ->
+        pr_update ppf upd_table upd_fields upd_where
+    | Insert { ins_table; ins_fields; ins_records } ->
+        pr_insert ppf ins_table ins_fields ins_records
+    | With (_, q, z, q') ->
+        match q' with
+        | Select (fields, tables, condition, os) ->
+            (* Inline the query *)
+            let tables = List.map (fun (t, x) -> Format.asprintf "%a as %s" pp_quote t (string_of_table_var x)) tables in
+            let pr_q ppf q = pr_query ppf quote ignore_fields q in
+            let q = Format.asprintf "(%s) as %s" (Format.asprintf "%a" pr_q q) (string_of_table_var z) in
+            pr_select ppf fields (q::tables) condition os
+        | _ -> assert false
 
-and string_of_base quote one_table b =
-  let sb = string_of_base quote one_table in
+and pr_base ppf quote one_table b =
+  let string_of_projection quote one_table (var, label) =
+  if one_table then
+    quote label
+  else
+    buffer_concat [string_of_table_var var; "."; (quote label)]
+  in
+  let pr_b ppf b = pr_base ppf quote one_table b in
+  let pr_q ppf q = pr_query ppf quote true q in
+  let unary_ops = StringSet.of_list ["intToString"; "stringToInt"; "intToFloat"; "floatToString"; "stringToFloat"; "floatToInt"; "not"; "negate"; "negatef"] in
+  let binary_ops = StringSet.of_list ["&&"; "||"; "=="; "<>"; "<"; ">"; "<="; ">="; "RLIKE"; "LIKE"] in
+  let binary_map op =
+    match op with
+      | "&&" -> "and"
+      | "||" -> "or"
+      | "==" -> "="
+      | _ -> op
+  in
+  let unary_map op =
+    match op with
+      | "floatToInt" -> "floor"
+      | "not" -> "not "
+      | "negate" | "negatef" -> "-"
+      | _ -> ""
+  in
     match b with
-      | Case (c, t, e) ->
-          "case when " ^ sb c ^ " then " ^sb t ^ " else "^ sb e ^ " end"
-      | Constant c -> Constant.to_string c
-      | Project (var, label) -> string_of_projection quote one_table (var, label)
-      | Apply (op, [l; r]) when Arithmetic.is op
-          -> Arithmetic.gen (sb l, op, sb r)
-      | Apply (("intToString" | "stringToInt" | "intToFloat" | "floatToString"
-                | "stringToFloat"), [v]) -> sb v
-      | Apply ("floatToInt", [v]) -> "floor("^sb v^")"
+      | Case (c, t, e) -> Format.fprintf ppf "case when %a then %a else %a end" pr_b c pr_b t pr_b e
+      | Constant c -> Format.fprintf ppf "%a" pp_constant (Constant.to_string c)
+      | Project (var, label) -> Format.fprintf ppf "%a" pp_constant (string_of_projection quote one_table (var, label))
+      | Apply (op, [l; r]) when Arithmetic.is op ->
+        let pr_b ppf b = pr_base ppf quote one_table b in
+        let l = Format.asprintf "%a" pr_b l in
```

Comment:
Fine for now but asprintf should eventually be avoided

Reply:
`Arithmetic.gen` requires `string` as arguments. I think `asprintf` is necessary to generate intermediate results here?